src/mbi/relaxed_projection.py

```diff
@@ -0,0 +1,368 @@
+        einsum_str = f'{input_dims}->a{output_dims}'
+
+        joint = jnp.einsum(einsum_str, *splits).mean(axis=0)
+        return np.array(joint).flatten()
+
+    def project(self, cl: tuple):
+        cols = list(itertools.chain.from_iterable(self.feature_indices_map[attr] for attr in cl))
+        D_proj = self.D_prime[:, cols]
+        domain_proj = self.domain.project(cl)
+        return ProjectableData(D_proj, domain_proj)
+    
+    def synthetic_data(self, num_samples, seed=0):
+        n, _ = self.D_prime.shape
+        key = jax.random.PRNGKey(seed)
+        key, sk = jax.random.split(key)
+        sampled_indices = jax.random.choice(key, n, shape=(num_samples,), replace=True)
+        sampled_data = self.D_prime[sampled_indices]
+
+        result = []
+        for attr in self.domain.attrs:
+            p = sampled_data[:, self.feature_indices_map[attr]]
+            key, sk = jax.random.split(key)
+            ids = jax.random.categorical(sk, jnp.log(jnp.clip(p, a_min=1e-30, a_max=1.0)), axis=-1)
+            result.append(ids)
+
+        sampled_df = pd.DataFrame(np.stack([np.array(r) for r in result], axis=1), columns = list(self.domain.attrs))
+        sampled_data = Dataset(sampled_df, self.domain)
+        return sampled_data
+
+
+def relaxed_projection_estimation(
+    domain: Domain,
+    measurements: List[tuple],
+    iters: int = 1000,
+    lr: float = 0.01,
+    D_start: jax.Array = None,
+    seed: int = 0,
+    D_size: int = 1000,
+) -> ProjectableData:
+    """
+    API function for relaxed projection mechanism, return a 
+    ProjectableData class, which can further be used to conduct 
+    projection and synthesize data.
+
+    Args:
+        domain: a Domain class.
+        measurements: a list of tuples, each of which should be in 
+                      form ``(clique, measurement of clique)``.
+        iters: number of iterations in training.
+        lr: learning rate in training.
+        D_start: initial data on which we start the optimization. If 
+                 None, the optimization will start from a randomly 
+                 initialized data.
+        seed: random seed.
+        D_size: number of records in optimizable data.
+    """
+    key = jax.random.PRNGKey(seed)
+    if D_start is None:
+        D_start = _initialize_synthetic_dataset(key, num_generated_points=D_size, data_dimension=np.sum(domain.shape))
+
+    stat_dim = _obtain_dim(measurements = measurements)
+    statistics = [MarginalStatistics(domain, dim) for dim in stat_dim]
+    selected_query_index, measured_ans = _marginal_stat(measurements, stat_dim, statistics)
+
+    exact_fn, diff_fn = [None]*len(stat_dim), [None]*len(stat_dim)
+    for i in range(len(stat_dim)):
+        exact_fn[i] = statistics[i].get_exact_statistics_fn()
+        diff_fn[i]  = statistics[i].get_differentiable_statistics_fn()
+
+    @jax.jit
+    def progress_loss(D_prime, query_idx, target, _exact_fn=exact_fn):
```

Comment:
There seems to be some code redundancy between progress_loss and loss_fn, can you unify them somehow?

Reply:
We also have callbacks.py for metrics logging purposes, could that be used here?  I'm not 100% sure, it might have some PGM-specific assumptions baked in, can't remember